[Blackjack - Step1] Submit Arekiu's mission

arekiu opened the pull request:
# Blackjack game implementation

## Features
- Create deck with 52 shuffled cards
- Use interface to create Dealer and Players (with custom names)
- Assign cards and count total value of cards in the hand
- Determine winners depending on their status as active in the game 

## Implementation notes
- My team reached the deadline, so the priority was to achieve the functionality in the game. Definitely I´m willing to improve different points, that I already see are not the best implementation:
    - Game() has methods that should belong to a separate Class
    - The use of magic numbers
    - Validation of input
    - Less use of "forEach"
    - Take more advantage of functional programming
- I would like to know if the interface was well implemented or if in this case, it´s better to use an Abstract class

Thank you so much for your time, and I hope our code is easy to follow!

alexisaguilera-dh committed:
feat: copy project from peer repository

Hyeon9mak commented on README.md:
```diff
@@ -1 +1,48 @@
+A simple kotlin project implementation of the classic Blackjack card game where a player competes against a dealer. 
+The goal is to get as close to 21 as possible without going over.
+
+## Features 
+- Input View 
+- [x] ask user inputs for player name
+- [x] get list of names from the user input
+
+- Card 
+- [x] Visibility of the card implemented by flipCard() 
+- [x] Determine the value of each card
+- [x] Create Enum class - Ranks
+- [x] Create Class - Card
+- [x] Create Enum class - Suits
+
+- Deck
+- [x] prepare deck
+- [x] shuffle cards
+
+- Participant interface 
+- [x] drawCard()
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/controller/Game.kt:
```diff
@@ -0,0 +1,70 @@
+                it.drawCard(deck.giveCard())
+            }
+        }
+    }
```
Right now, it looks like the Dealer and Player objects are just exchanging properties by exposing them to each other.

How about letting the Dealer and Player interact with each other more directly instead?

This would reduce the amount of information exposed to the Controller, making the code easier to maintain — and it would likely make writing test code much easier too!

```kotlin
dealer.selfDrawInitCards()
dealer.giveInitCardsToPlayers(players = players)
```

arekiu replied:
Great idea! In order to improve the Game logic, I added the Deck as Dealer private property.

Before changing this logic, Players were accessing directly to the Deck and the responsibilities of Dealer were not clear.

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,15 @@
+package blackjack.model
+
+class Dealer(
+    override val name: String = "Dealer",
```
If the name "Dealer" needs to be changed to below image file, which class file will a fellow developer open first: `OutputView`-`Dealer`?

<img width="862" height="534" alt="image" src="https://github.com/user-attachments/assets/f0acb4f5-574d-46e2-a121-ac3b36268fb5" />

arekiu replied:
If I understood right your question, you mean for example this:

`fun displayDealerDrawMessage() {
        println("Dealer draws one more card due to having 16 or less.")
    }`
    
   because I used the word "Dealer" multiple times in OutputView and then if I modify the name (like "CardGiver" or "Groupier") I need to make multiple changes in the file.
   
   I replaced for this:
   
   `fun displayDealerDrawMessage(dealer: Dealer) {
        println("${dealer.name} draws one more card due to having 16 or less.\n")
    }`
    
Please let me know in case I misunderstood the request.

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,15 @@
+
+class Dealer(
+    override val name: String = "Dealer",
+    override var isActive: Boolean = true,
```
What does `isActive` mean? 🤔 
I think it would be more intuitive to use actual blackjack rules!

arekiu replied:
Clear! Replaced for "isBusted"

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,15 @@
+        val lose = players.count { it.isActive }
+        val win = players.count { !it.isActive }
+        return "Dealer: $win Win $lose Lose"
+    }
```
It looks like View's responsibility. How do u think?

arekiu replied:
Absolutely it belongs to OutputView

Hyeon9mak commented on src/main/kotlin/blackjack/model/Deck.kt:
```diff
@@ -0,0 +1,19 @@
+    fun giveCard(): Card {
+        require(cards.isNotEmpty()) { "Deck is empty!" }
+        return cards.removeAt(0)
+    }
```
```suggestion
    fun giveCard(): Card {
        require(cards.isNotEmpty()) { "Deck is empty!" }
        return cards.removeFirst()
    }
```

😉

arekiu replied:
Thank you, I didn´t know that method!

Hyeon9mak commented on src/main/kotlin/blackjack/model/Participant.kt:
```diff
@@ -0,0 +1,37 @@
+package blackjack.model
+
+interface Participant {
+    val name: String
+    var isActive: Boolean
+    val cardsInHand: MutableList<Card>
```
What is the different between interface and abstract class?

arekiu replied:
I have been thinking about it. 
Interface is better when I want to define a role (behaviour by methods). Also if the class is going to have different roles, it could inherit from different interfaces. 
Abstract class is better to create classes that share properties and methods. So I think that it´s the best option for my project. 

Thank you for the question to making me reflect!

Hyeon9mak commented on src/main/kotlin/blackjack/model/Suit.kt:
```diff
@@ -0,0 +1,8 @@
+package blackjack.model
+
+enum class Suit(val symbol: Char) {
+    HEARTS('♥'),
+    DIAMONDS('♦'),
+    CLUBS('♣'),
+    SPADES('♠'),
+}
```
When the Blackjack app convert to Web browser image game,
fellow develop should check the `InputView`, `OutputView`, `Card`, ... a lot of classes.

How about move all view logic to View classes?
Or, u can make duplicate enum class in View class file only used for output on terminal.

```kotlin
enum class SuitView(
    val suit: Suit,
    val symbol: String,
) {
    HEART(suit = Suit.HEART, symbol = "♥"),
    // ...
}
```

arekiu replied:
Oh I never thought about it. But it has sense because, for the logic of the game, I don´t need to know the symbols and face at all.
I created another enum class for the FaceView as well.

Hyeon9mak commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    fun askToDrawCard(name: String): Boolean {
+        println("Would $name like to draw another card? (y for yes, n for no)")
+        val choice = readln()
+        return when (choice) {
+            "y", "Y" -> true
+            "n", "N" -> false
+            else -> throw IllegalArgumentException("Input can only contain y for yes, n for no")
+        }
+    }
```
"y", "n" strings are not important things of Blackjack game.
Those are just user's input buttons. 💯 👍

Or u can try this

```kotlin
package blackjack.view 

enum class UserAnswer(
    private val value: String,
) {
    YES(value = "y"),
    NO(value = "n"),
    ;

    companion object {
        fun findByValue(value: String): UserAnswer = entries.find { value.equals(other = it.value, ignoreCase = true) }
            ?: throw IllegalArgumentException("Invalid user answer: $value")
    }
}
```

arekiu replied:
I appreciate a lot this suggestion and I was analyzing about the benefits of implement it or not.
Finally, I consider that it would add more complexity in this case where the input is simple, and as you mention not an important part of the game. I have the feeling that currently, that block using "when" is readable and simple.
 But definitively, I would apply this change if the functionality of the game depends on different options.

Of course, if you ask me to implement the change, I´m open and willing to do it!

Hyeon9mak commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,58 @@
+import blackjack.model.Player
+import kotlin.collections.joinToString
+
+class OutputView {
```
There is no property.
So u can change this class to object class!

arekiu replied:
That´s right! I did the same with InputView

Hyeon9mak commented on src/main/kotlin/blackjack/model/Participant.kt:
```diff
@@ -0,0 +1,37 @@
+        return totalValueOfCards
+    }
+
+    fun updateActiveStatus(totalValueOfCards: Int) {
+        if (totalValueOfCards > 21) {
+            isActive = false
+        }
+    }
```
I think u can give a new name to `21` score 😄

arekiu replied:
I created an object to store Constant values, now that 21 is "HAND_VALUE_LIMIT"

Hyeon9mak reviewed (changes requested):
Hello Alex!
U did a good job to complete Blackjack mission step1 👍

It looks like time was a bit tight, but I can see you made an effort to follow the basic MVC pattern.
It would be great to polish up the MVC structure a bit more.

Also, I noticed there are many parts in the code that could be replaced with actual blackjack game rules and terminology from the real world.
I think using real-world concepts would make the code much easier for all developers to understand. 
It would be awesome if you could give this some thought! 😄

Feel free to ask any questions u have!

alexisaguilera-dh committed:
refactor(Participant): convert to abstract class

make Player and Dealer class to inherit from Participant  abstract class

alexisaguilera-dh committed:
refactor(Deck): replace list method to remove first element

alexisaguilera-dh committed:
refactor(Participant): modify constructor to assign name

alexisaguilera-dh committed:
refactor(Card): add method showCardFace()

alexisaguilera-dh committed:
refactor(OutputView): replace Player/Dealer parameters with Participant for polymorphism

alexisaguilera-dh committed:
refactor(View): create Input and Output View as object instead of class

alexisaguilera-dh committed:
refactor(OutputView): replace hardcoded value

alexisaguilera-dh committed:
refactor(OutputView): move method from Dealer class

alexisaguilera-dh committed:
refactor(Dealer): give Deck property and methods

add methods to interact directly with Deck and Player classes

alexisaguilera-dh committed:
feat(BlackJAckValues): create object to store constant values

alexisaguilera-dh committed:
refactor: use BlackJack rules for naming

alexisaguilera-dh committed:
refactor(View): create enum classes to display card´s rank and face

alexisaguilera-dh committed:
refactor(Cards): create wrapper class to hold cards methods

alexisaguilera-dh committed:
refactor: format files

alexisaguilera-dh committed:
refactor(Participant): check condition from internal property

alexisaguilera-dh committed:
refactor(Participant): add private setter for isBusted

alexisaguilera-dh committed:
refactor(Game): check if Dealer is Boosted at the end of game

alexisaguilera-dh committed:
refactor(tests): increase test coverage

alexisaguilera-dh committed:
refactor(Card): remove unnecesary method

arekiu commented:
Hello GuMak!

I implemented all the requested changes (except the one in Inputview but I left my comment there for your consideration).
I repeat, if you consider that I should apply that change, please let me know, because my wish is to learn about better practices!

Also, I refactored extra things to make the project cleaner:
- Created an object to store constant values
- Create wrapper class for Cards
- Increased test coverage
- Divided methods depending on Class responsibility

Thank you so much for your time and your attention to review my code.
Have a nice day!

Hyeon9mak commented on src/main/kotlin/blackjack/controller/Game.kt:
```diff
@@ -0,0 +1,67 @@
+        compareFinalCards(players)
+    }
+
+    private fun assignInitialCards(players: List<Player>) {
+        dealer.selfDrawInitialCards()
+        dealer.dealInitialCardsToPlayers(players)
+        OutputView.displayInitialCards(dealer, players)
+    }
+
+    private fun hitOrStay(players: List<Player>) {
+        askPlayersToHit(players)
+        dealerDraws()
+        OutputView.displayCardsWithTotalValue(dealer)
+        players.forEach { player -> OutputView.displayCardsWithTotalValue(player) }
+    }
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,27 @@
+package blackjack.model
+
+class Dealer(name: String = "Dealer") : Participant(name) {
+    private val deck = Deck()
+
+    fun mustDraw(totalValueOfCards: Int) = totalValueOfCards < BlackJackValues.DEALER_STAND_CONDITION
+
+    fun selfDrawCard() {
```
> Please let me know in case I misunderstood the request.

Oh, what I meant was "Really need to give the dealer a name?"
In a Blackjack, the dealer's name doesn't seem to be an important part of the business logic.
It's more of a UI-level distinction between the player and the dealer.

arekiu replied:
Now I understand your point! I could create the Dealer without assigning the name.
Anyway, I would keep the field because if for some reason, someone wants to be the Dealer, I can replace the name while creating the Object.
In the other hand, for myself, is more visually clear when I see $dealer.name in a string, indicating from who is coming that action. I guess another option would be creating a method to show the name of the class itself?

Hyeon9mak commented on src/main/kotlin/blackjack/controller/Game.kt:
```diff
@@ -0,0 +1,67 @@
+                    dealer.dealCardToPlayer(it)
+                    it.checkCardsValueLimit()
+                    println(OutputView.showHandCards(it, false))
+                }
+            } while (answer && !it.isBusted)
+        }
```
And I think u can remove the `!` operator by negative question method.

```kotlin
while (answer && it.isNotBusted)
```

arekiu replied:
Now I replaced isBusted for isPlaying in Participant class. I think is more clear because a player could go out of the game if they are busted but also if they have less points than dealer.

Hyeon9mak commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -0,0 +1,4 @@
+package blackjack.model
+
+class Card(val rank: Rank, val suit: Suit) {
+}
```
Question.
one `♦️A` card and another `♦️A` card on the table.

Should we treat these as the same card or as two distinct cards?

That decision could help determine whether we should use a data class or a regular class here.

arekiu replied:
I understand. As class they would be different because their equality is based in reference. But as data class, their equality is based in their attributes. So it´s better for Card, cause they just store immutable values

Hyeon9mak commented on src/main/kotlin/blackjack/model/Participant.kt:
```diff
@@ -0,0 +1,23 @@
+package blackjack.model
+
+abstract class Participant(
+    val name: String,
+    val cardsInHand: Cards = Cards(mutableListOf())
+) {
+    var isBusted: Boolean = false
+        private set
+
+    fun drawCard(card: Card) {
+        cardsInHand.addCard(card)
+    }
+
+    fun updateBustedStatus() {
+            isBusted = true
+    }
+
```
Managing the busted status as a property is a valid approach,
but we could also consider asking the Hand in real-time whenever we need that information.

```kotlin
cardInHand.isBusted()
```

arekiu replied:
Ok clear!
I added this method in Cards class: fun checkIfIsBustHand()

Hyeon9mak commented on src/main/kotlin/blackjack/view/RankView.kt:
```diff
@@ -0,0 +1,23 @@
+
+import blackjack.model.Rank
+
+enum class RankView(val rank: Rank, val face: String) {
+    TWO(Rank.TWO, "2"),
+    THREE(Rank.THREE, "3"),
+    FOUR(Rank.FOUR, "4"),
+    FIVE(Rank.FIVE, "5"),
+    SIX(Rank.SIX, "6"),
+    SEVEN(Rank.SEVEN, "7"),
+    EIGHT(Rank.EIGHT, "8"),
+    NINE(Rank.NINE, "9"),
+    TEN(Rank.TEN, "10"),
+    JACK(Rank.JACK, "J"),
+    QUEEN(Rank.QUEEN, "Q"),
+    KING(Rank.KING, "K"),
+    ACE(Rank.ACE, "A");
```
😉

Hyeon9mak reviewed (approved):
Hello Alex!
Great job the reflect the feedback.

I’ve left a few minor comments.
feel free to keep them in mind as you move on to the step2.
And if you have any questions, don’t hesitate to ask!

Hope u have a nice weekend!
